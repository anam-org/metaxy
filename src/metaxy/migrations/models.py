"""Data models for migration system."""

from datetime import datetime
from typing import TYPE_CHECKING, Any, Literal

import pydantic

if TYPE_CHECKING:
    from metaxy.migrations.ops import BaseOperation


def _load_operation_class(class_path: str) -> type:
    """Dynamically import operation class.

    Args:
        class_path: Full class path (e.g., "myproject.migrations.S3Backfill")

    Returns:
        Operation class

    Raises:
        ImportError: If module or class not found
    """
    from metaxy.migrations.ops import BaseOperation

    module_path, class_name = class_path.rsplit(".", 1)
    module = __import__(module_path, fromlist=[class_name])
    cls = getattr(module, class_name)

    if not issubclass(cls, BaseOperation):
        raise TypeError(f"{class_path} must be a subclass of BaseOperation, got {cls}")

    return cls


def _register_operation_classes(operations_data: list[dict[str, Any]]) -> None:
    """Register all operation classes for Pydantic discriminated union.

    Dynamically imports user-defined operation classes so Pydantic can
    deserialize them based on the 'type' discriminator field.

    Args:
        operations_data: List of operation dicts from YAML
    """
    for op_data in operations_data:
        op_type = op_data.get("type")
        if not op_type:
            continue

        # Try to import the class (registers it with Pydantic)
        try:
            _load_operation_class(op_type)
        except Exception:
            # Will fail later during validation if class doesn't exist
            # This is just best-effort registration
            pass


class Migration(pydantic.BaseModel):
    """Complete migration definition.

    A migration contains a list of operations that are executed in order.
    Each operation can be any subclass of BaseOperation (e.g.,
    DataVersionReconciliation, MetadataBackfill, or user-defined).

    Operations are polymorphic - deserialized based on the 'type' field.

    Migrations form an explicit dependency chain via parent_migration_id.
    This ensures migrations are applied in the correct order without
    relying on implicit timestamp-based sorting.
    """

    version: int  # Migration schema version
    id: str  # Unique ID (format: "migration_YYYYMMDD_HHMMSS")
    parent_migration_id: str | None = (
        None  # Parent migration (None for first migration)
    )
    from_snapshot_id: (
        str  # Feature graph snapshot before migration (current state in store)
    )
    to_snapshot_id: str  # Feature graph snapshot after migration (target state)
    description: str
    created_at: datetime  # When migration was created

    # Optional overrides for moved/renamed feature classes
    feature_class_overrides: dict[str, str] = pydantic.Field(default_factory=dict)

    # Operations as list of dicts - will be validated/loaded dynamically
    # Can't use Union with discriminator because we support user-defined subclasses
    operations: list[dict[str, Any]]

    def get_operations(self) -> list["BaseOperation"]:
        """Parse operations into BaseOperation objects.

        Returns:
            List of parsed operation objects

        Raises:
            ImportError: If operation type cannot be imported
            ValidationError: If operation data is invalid
        """

        parsed_operations = []

        for op_data in self.operations:
            op_type = op_data.get("type")
            if not op_type:
                raise ValueError(f"Operation missing 'type' field: {op_data}")

            # Load the operation class
            op_class = _load_operation_class(op_type)

            # Parse and validate
            op_instance = op_class.model_validate(op_data)  # type: ignore[attr-defined]
            parsed_operations.append(op_instance)

        return parsed_operations

    @staticmethod
    def from_yaml(path: str) -> "Migration":
        """Load migration from YAML file.

        Supports polymorphic operations - dynamically loads operation classes
        based on the 'type' field.

        Args:
            path: Path to YAML migration file

        Returns:
            Migration object

        Raises:
            FileNotFoundError: If file doesn't exist
            pydantic.ValidationError: If YAML structure is invalid
            ImportError: If operation type class cannot be imported
        """
        import yaml

        with open(path) as f:
            data = yaml.safe_load(f)

        # Register operation classes for polymorphic deserialization
        _register_operation_classes(data.get("operations", []))

        # Pydantic will handle validation and parsing with discriminated unions
        return Migration.model_validate(data)

    def to_yaml(self, path: str) -> None:
        """Save migration to YAML file.

        Args:
            path: Path to write YAML file
        """
        import yaml

        # Use Pydantic's model_dump for serialization
        # Use by_alias=True to export "from" instead of "from_"
        data = self.model_dump(mode="python", by_alias=True)

        with open(path, "w") as f:
            yaml.dump(data, f, default_flow_style=False, sort_keys=False)


class MigrationResult(pydantic.BaseModel):
    """Result of executing a migration."""

    migration_id: str
    status: Literal["completed", "failed", "skipped"]
    operations_applied: int
    operations_failed: int
    affected_features: list[str]
    errors: dict[str, str]  # feature_key -> error message
    duration_seconds: float
    timestamp: datetime

    def summary(self) -> str:
        """Human-readable summary of migration result.

        Returns:
            Multi-line summary string
        """
        lines = [
            f"Migration: {self.migration_id}",
            f"Status: {self.status.upper()}",
            f"Timestamp: {self.timestamp.isoformat()}",
            f"Duration: {self.duration_seconds:.2f}s",
            f"Operations: {self.operations_applied} applied, {self.operations_failed} failed",
            f"Affected features: {len(self.affected_features)}",
        ]

        if self.affected_features:
            lines.append("\nFeatures updated:")
            for feature in self.affected_features:
                lines.append(f"  âœ“ {feature}")

        if self.errors:
            lines.append("\nErrors:")
            for feature, error in self.errors.items():
                lines.append(f"  âœ— {feature}: {error}")

        return "\n".join(lines)
