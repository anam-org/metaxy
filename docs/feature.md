## Feature Version

Every feature has a **feature version** - hash of its complete specification:
- Feature key
- Field definitions and code versions
- Dependencies (feature-level and field-level)

Being more pedantic, the feature version is calculated by hashing together its field versions and the feature key. The field version is calculated by hashing together the field code version, key, and versions of its dependencies.

Feature and field versions are static and deterministic. They are available as class methods:

```python
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=1),
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

# Get feature version
print(VideoProcessing.feature_version())  # "a3f8b2c1"

# Change code_version
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=2),  # Changed!
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

print(VideoProcessing.feature_version())  # "d7e9f4a2" (different!)
```

Every metadata row includes a `feature_version` column:
```python
metadata = pl.DataFrame({
    "sample_id": [1, 2, 3],
    "data_version": [...],
    "feature_version": "a3f8b2c1",  # Auto-added by store
})
```

This enables:
- **Current vs historical**: Filter by `feature_version` to get specific versions
- **Migration tracking**: Know which rows need updating
- **Audit trail**: See what feature definition produced each row


## User-Defined Metadata Alignment

When feature definitions change, you may need custom logic to align metadata with upstream changes. Override `Feature.align_metadata_with_upstream`:

### Default Behavior (Inner Join)

```python
class MyFeature(Feature, spec=...):
    # Default behavior: inner join on sample_id
    # Only keeps samples present in ALL upstream features
    pass

# During migration:
# - Reads current metadata
# - Inner joins with upstream on sample_id
# - Only samples present in both current AND upstream survive
```

### Custom Alignment Examples

#### One-to-Many Mapping

```python
class VideoFrames(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "frames"]),
    deps=[FeatureDep(key=FeatureKey(["videos"]))],
    ...
)):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Each video produces 30 frames.

        When videos upstream change, regenerate all frame sample IDs.
        """
        video_samples = upstream_metadata["videos"]["sample_id"]

        # Generate frame sample IDs
        frames = []
        for video_id in video_samples:
            for frame_idx in range(30):
                frames.append({
                    "sample_id": f"{video_id}_frame_{frame_idx}",
                    "video_id": video_id,
                    "frame_idx": frame_idx,
                })

        return pl.DataFrame(frames)
```

#### Conditional Filtering

```python
class ProcessedVideos(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Only process videos longer than 10 seconds."""

        videos = upstream_metadata["videos"]

        # Filter by duration
        valid_videos = videos.filter(pl.col("duration") > 10)

        # Keep custom columns from current_metadata if available
        if len(current_metadata) > 0:
            return current_metadata.join(
                valid_videos.select(pl.col("sample_id")),
                on="sample_id",
                how="inner",  # Only keep samples that pass filter
            )
        else:
            return valid_videos.select(["sample_id", "duration", "path"])
```

#### Outer Join (Keep All Samples)

```python
class MergedFeature(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Keep union of all upstream samples (outer join)."""

        all_samples = set()
        for upstream_df in upstream_metadata.values():
            all_samples.update(upstream_df["sample_id"].to_list())

        return pl.DataFrame({"sample_id": sorted(all_samples)})
```
