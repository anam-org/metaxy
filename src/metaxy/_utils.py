import narwhals as nw
import polars as pl


def collect_to_polars(lazy_frame: nw.LazyFrame) -> pl.DataFrame:
    """Helper to collect a Narwhals LazyFrame and convert to Polars DataFrame.

    This handles all backend conversions (Polars, DuckDB/PyArrow, etc.) transparently.
    Used throughout tests for materializing query results.
    """
    return lazy_frame.collect().to_polars()
