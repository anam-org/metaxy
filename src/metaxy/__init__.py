from metaxy.models.container import ContainerDep, ContainerSpec, SpecialContainerDep
from metaxy.models.feature import Feature
from metaxy.models.feature_spec import FeatureDep, FeatureDepMetadata, FeatureSpec
from metaxy.models.types import ContainerKey, FeatureKey

__all__ = [
    "Feature",
    "FeatureDep",
    "FeatureDepMetadata",
    "FeatureSpec",
    "ContainerDep",
    "ContainerSpec",
    "SpecialContainerDep",
    "FeatureKey",
    "ContainerKey",
]
