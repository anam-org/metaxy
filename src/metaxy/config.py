"""Configuration system for Metaxy using pydantic-settings."""

from pathlib import Path
from typing import TYPE_CHECKING, Any, TypeVar, overload

try:
    import tomllib  # Python 3.11+
except ImportError:
    import tomli as tomllib  # Fallback for Python 3.10

from pydantic import Field as PydanticField
from pydantic_settings import (
    BaseSettings,
    PydanticBaseSettingsSource,
    SettingsConfigDict,
)

if TYPE_CHECKING:
    from metaxy.metadata_store.base import MetadataStore
    from metaxy.models.feature import FeatureRegistry

T = TypeVar("T")


def load_registry(registry_path: str) -> "FeatureRegistry":
    """Load registry from import path.

    Args:
        registry_path: Import path like "myapp.features:my_registry"

    Returns:
        FeatureRegistry instance

    Raises:
        ValueError: If path is invalid
        ImportError: If module or registry not found
        TypeError: If loaded object is not a FeatureRegistry
    """
    from metaxy.models.feature import FeatureRegistry

    if ":" not in registry_path:
        raise ValueError(
            f"Invalid registry path: {registry_path}. "
            f"Expected format: 'module.path:registry_name'"
        )

    module_path, registry_name = registry_path.split(":", 1)

    try:
        module = __import__(module_path, fromlist=[registry_name])
        registry = getattr(module, registry_name)
    except (ImportError, AttributeError) as e:
        raise ImportError(f"Failed to load registry from {registry_path}: {e}") from e

    if not isinstance(registry, FeatureRegistry):
        raise TypeError(
            f"Registry at {registry_path} is not a FeatureRegistry instance. "
            f"Got: {type(registry)}"
        )

    return registry


class TomlConfigSettingsSource(PydanticBaseSettingsSource):
    """Custom settings source for TOML configuration files.

    Auto-discovers configuration in this order:
    1. Explicit file path if provided
    2. metaxy.toml in current directory (preferred)
    3. pyproject.toml [tool.metaxy] section (fallback)
    4. No config (returns empty dict)
    """

    def __init__(self, settings_cls: type[BaseSettings], toml_file: Path | None = None):
        super().__init__(settings_cls)
        self.toml_file = toml_file or self._discover_config_file()
        self.toml_data = self._load_toml()

    def _discover_config_file(self) -> Path | None:
        """Auto-discover config file."""
        # Prefer metaxy.toml
        if Path("metaxy.toml").exists():
            return Path("metaxy.toml")

        # Fallback to pyproject.toml
        if Path("pyproject.toml").exists():
            return Path("pyproject.toml")

        return None

    def _load_toml(self) -> dict[str, Any]:
        """Load TOML file and extract metaxy config."""
        if self.toml_file is None:
            return {}

        with open(self.toml_file, "rb") as f:
            data = tomllib.load(f)

        # Extract [tool.metaxy] from pyproject.toml or root from metaxy.toml
        if self.toml_file.name == "pyproject.toml":
            return data.get("tool", {}).get("metaxy", {})
        else:
            return data

    def get_field_value(self, field: Any, field_name: str) -> tuple[Any, str, bool]:
        """Get field value from TOML data."""
        field_value = self.toml_data.get(field_name)
        return field_value, field_name, False

    def __call__(self) -> dict[str, Any]:
        """Return all settings from TOML."""
        return self.toml_data


class StoreConfig(BaseSettings):
    """Configuration for a single metadata store.

    Structure:
        type: Full import path to store class
        config: Dict of all configuration (including fallback_stores)

    Example:
        >>> config = StoreConfig(
        ...     type="metaxy_delta.DeltaMetadataStore",
        ...     config={
        ...         "table_uri": "s3://bucket/metadata",
        ...         "region": "us-west-2",
        ...         "fallback_stores": ["prod"],
        ...     }
        ... )
    """

    model_config = SettingsConfigDict(
        extra="forbid",  # Only type and config fields allowed
    )

    # Store class (full import path)
    type: str

    # Store configuration (all kwargs for __init__)
    # This includes fallback_stores, table_uri, db_path, storage_options, etc.
    config: dict[str, Any] = PydanticField(default_factory=dict)


class MetaxyConfig(BaseSettings):
    """Main Metaxy configuration.

    Loads from:
    1. TOML file (metaxy.toml or pyproject.toml [tool.metaxy])
    2. Environment variables (METAXY_*)
    3. Init arguments

    Priority: init > env vars > TOML

    Example:
        >>> # Auto-discover config
        >>> config = MetaxyConfig.load()
        >>>
        >>> # Get store instance
        >>> store = config.get_store("prod")
        >>>
        >>> # Override via env var
        >>> # METAXY_STORE=staging METAXY_REGISTRY=myapp.features:my_registry
        >>> config = MetaxyConfig.load()
        >>> store = config.get_store()  # Uses staging with custom registry
    """

    model_config = SettingsConfigDict(
        env_prefix="METAXY_",
        env_nested_delimiter="__",
    )

    # Store to use
    store: str = "dev"

    # Named store configurations
    stores: dict[str, StoreConfig] = PydanticField(default_factory=dict)

    # Migrations directory
    migrations_dir: str = "metaxy/migrations"

    # Entrypoints to load (list of module paths)
    entrypoints: list[str] = PydanticField(default_factory=list)

    @classmethod
    def settings_customise_sources(
        cls,
        settings_cls: type[BaseSettings],
        init_settings: PydanticBaseSettingsSource,
        env_settings: PydanticBaseSettingsSource,
        dotenv_settings: PydanticBaseSettingsSource,
        file_secret_settings: PydanticBaseSettingsSource,
    ) -> tuple[PydanticBaseSettingsSource, ...]:
        """Customize settings sources: init → env → TOML.

        Priority (first wins):
        1. Init arguments
        2. Environment variables
        3. TOML file
        """
        toml_settings = TomlConfigSettingsSource(settings_cls)
        return (init_settings, env_settings, toml_settings)

    @classmethod
    def load(
        cls, config_file: str | Path | None = None, *, search_parents: bool = True
    ) -> "MetaxyConfig":
        """Load config with auto-discovery and parent directory search.

        Args:
            config_file: Optional config file path (overrides auto-discovery)
            search_parents: Search parent directories for config file (default: True)

        Returns:
            Loaded config (TOML + env vars merged)

        Example:
            >>> # Auto-discover with parent search
            >>> config = MetaxyConfig.load()
            >>>
            >>> # Explicit file
            >>> config = MetaxyConfig.load("custom.toml")

            >>> # Auto-discover without parent search
            >>> config = MetaxyConfig.load(search_parents=False)
        """
        # Search for config file if not explicitly provided
        if config_file is None and search_parents:
            config_file = cls._discover_config_with_parents()

        # For explicit file, temporarily patch the TomlConfigSettingsSource
        # to use that file, then use normal instantiation
        # This ensures env vars still work

        if config_file:
            # Create a custom settings source class for this file
            toml_path = Path(config_file)

            class CustomTomlSource(TomlConfigSettingsSource):
                def __init__(self, settings_cls: type[BaseSettings]):
                    # Skip auto-discovery, use explicit file
                    super(TomlConfigSettingsSource, self).__init__(settings_cls)
                    self.toml_file = toml_path
                    self.toml_data = self._load_toml()

            # Customize sources to use custom TOML file
            original_method = cls.settings_customise_sources

            @classmethod  # type: ignore[misc]
            def custom_sources(
                cls_inner,
                settings_cls,
                init_settings,
                env_settings,
                dotenv_settings,
                file_secret_settings,
            ):
                toml_settings = CustomTomlSource(settings_cls)
                return (init_settings, env_settings, toml_settings)

            # Temporarily replace method
            cls.settings_customise_sources = custom_sources  # type: ignore[assignment]
            config = cls()
            cls.settings_customise_sources = original_method  # type: ignore[method-assign]
        else:
            # Use default sources (auto-discovery + env vars)
            config = cls()

        # Load entrypoints if configured (do this after config is created)
        if config.entrypoints:
            from metaxy.entrypoints import load_config_entrypoints

            load_config_entrypoints(config.entrypoints)

        return config

    @staticmethod
    def _discover_config_with_parents() -> Path | None:
        """Discover config file by searching current and parent directories.

        Searches for metaxy.toml or pyproject.toml in current directory,
        then iteratively searches parent directories.

        Returns:
            Path to config file if found, None otherwise
        """
        current = Path.cwd()

        while True:
            # Check for metaxy.toml (preferred)
            metaxy_toml = current / "metaxy.toml"
            if metaxy_toml.exists():
                return metaxy_toml

            # Check for pyproject.toml
            pyproject_toml = current / "pyproject.toml"
            if pyproject_toml.exists():
                return pyproject_toml

            # Move to parent
            parent = current.parent
            if parent == current:
                # Reached root
                break
            current = parent

        return None

    def get_store(
        self,
        name: str | None = None,
    ) -> "MetadataStore":
        """Instantiate metadata store by name.

        Args:
            name: Store name (uses config.store if None)

        Returns:
            Instantiated metadata store

        Raises:
            ValueError: If store name not found in config, or if fallback stores
                have different hash algorithms than the parent store
            ImportError: If store class cannot be imported

        Example:
            >>> config = MetaxyConfig.load()
            >>> store = config.get_store("prod")
            >>>
            >>> # Use default store
            >>> store = config.get_store()
        """
        from metaxy.data_versioning.hash_algorithms import HashAlgorithm

        if len(self.stores) == 0:
            raise ValueError(
                "No Metaxy stores available. They should be configured in metaxy.toml|pyproject.toml or via environment variables."
            )

        name = name or self.store

        if name not in self.stores:
            raise ValueError(
                f"Store '{name}' not found in config. "
                f"Available stores: {list(self.stores.keys())}"
            )

        store_config = self.stores[name]

        # Import store class
        store_class = self._import_class(store_config.type)

        # Extract configuration
        config_copy = store_config.config.copy()
        fallback_store_names = config_copy.pop("fallback_stores", [])

        # Get hash_algorithm from config (if specified) and convert to enum
        configured_hash_algorithm = config_copy.get("hash_algorithm")
        if configured_hash_algorithm is not None:
            # Convert string to enum if needed
            if isinstance(configured_hash_algorithm, str):
                configured_hash_algorithm = HashAlgorithm(configured_hash_algorithm)
                config_copy["hash_algorithm"] = configured_hash_algorithm
        else:
            # Use default
            configured_hash_algorithm = HashAlgorithm.XXHASH64
            config_copy["hash_algorithm"] = configured_hash_algorithm

        # Build fallback stores recursively
        fallback_stores = []
        for fallback_name in fallback_store_names:
            fallback_store = self.get_store(fallback_name)
            fallback_stores.append(fallback_store)

        # Instantiate store with config + fallback_stores
        store = store_class(
            fallback_stores=fallback_stores,
            **config_copy,
        )

        # Verify the store actually uses the hash algorithm we configured
        # (in case a store subclass overrides the default or ignores the parameter)
        if store.hash_algorithm != configured_hash_algorithm:
            raise ValueError(
                f"Store '{name}' ({store_class.__name__}) was configured with "
                f"hash_algorithm='{configured_hash_algorithm.value}' but is using "
                f"'{store.hash_algorithm.value}'. The store class may have overridden "
                f"the hash algorithm. All stores must use the same hash algorithm."
            )

        return store

    @staticmethod
    def _import_class(class_path: str) -> type:
        """Import class from module path.

        Args:
            class_path: Full import path like "metaxy.metadata_store.InMemoryMetadataStore"

        Returns:
            Imported class

        Raises:
            ImportError: If module or class not found
        """
        module_path, class_name = class_path.rsplit(".", 1)
        module = __import__(module_path, fromlist=[class_name])
        return getattr(module, class_name)

    @staticmethod
    @overload
    def _import_object(import_path: str, expected_type: type[T]) -> T: ...

    @staticmethod
    @overload
    def _import_object(import_path: str, expected_type: None = None) -> Any: ...

    @staticmethod
    def _import_object(
        import_path: str, expected_type: type[T] | None = None
    ) -> T | Any:
        """Import object from module:object path.

        Args:
            import_path: Import path like "module.path:object_name"
            expected_type: Optional type to validate against

        Returns:
            Imported object (typed as expected_type if provided)

        Raises:
            ValueError: If path format is invalid
            ImportError: If module or object not found
            TypeError: If object doesn't match expected_type
        """
        if ":" not in import_path:
            raise ValueError(
                f"Invalid import path: {import_path}. "
                f"Expected format: 'module.path:object_name'"
            )

        module_path, object_name = import_path.split(":", 1)

        try:
            module = __import__(module_path, fromlist=[object_name])
            obj = getattr(module, object_name)
        except (ImportError, AttributeError) as e:
            raise ImportError(f"Failed to load object from {import_path}: {e}") from e

        if expected_type is not None and not isinstance(obj, expected_type):
            type_name = getattr(expected_type, "__name__", str(expected_type))
            raise TypeError(
                f"Object at {import_path} is not a {type_name} instance. "
                f"Got: {type(obj)}"
            )

        return obj
