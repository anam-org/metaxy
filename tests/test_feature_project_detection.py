"""Tests for feature project detection and assignment.

This module tests the multi-project architecture, verifying that:
1. Features get correct projects from configuration
2. Project names are validated correctly
3. Features from different projects can coexist in the same graph
"""

from __future__ import annotations

from pathlib import Path

import pytest
from syrupy.assertion import SnapshotAssertion

from metaxy import Feature, FeatureKey, FeatureSpec, FieldKey, FieldSpec
from metaxy.config import MetaxyConfig
from metaxy.models.feature import FeatureGraph

# Note: Module-level test features will be defined dynamically in tests
# to ensure they get the correct project from the test's config


def test_feature_gets_project_from_config(snapshot: SnapshotAssertion) -> None:
    """Test that features get project from MetaxyConfig."""
    # Create config with specific project
    config = MetaxyConfig(project="test_project")
    MetaxyConfig.set(config)

    try:
        # Create feature - should get project from config
        class TestFeature(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["test", "feature"]),
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

        # Feature should have project from config
        assert TestFeature.project == "test_project"
        assert TestFeature.project == snapshot

    finally:
        MetaxyConfig.reset()


def test_feature_project_different_configs(snapshot: SnapshotAssertion) -> None:
    """Test that features in different graphs can have different projects."""
    # Graph 1: project_a
    config_a = MetaxyConfig(project="project_a")
    graph_a = FeatureGraph()

    MetaxyConfig.set(config_a)

    with graph_a.use():

        class FeatureA(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["shared", "feature"]),
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Graph 2: project_b
    config_b = MetaxyConfig(project="project_b")
    graph_b = FeatureGraph()

    MetaxyConfig.set(config_b)

    with graph_b.use():

        class FeatureB(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["shared", "feature"]),
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Features should have different projects
    assert FeatureA.project == "project_a"
    assert FeatureB.project == "project_b"
    assert {"feature_a": FeatureA.project, "feature_b": FeatureB.project} == snapshot

    MetaxyConfig.reset()


def test_project_name_validation_empty() -> None:
    """Test that empty project names are rejected."""
    with pytest.raises(ValueError, match="project name cannot be empty"):
        MetaxyConfig(project="")


def test_project_name_validation_forward_slash() -> None:
    """Test that project names cannot contain forward slashes."""
    with pytest.raises(
        ValueError, match="project name .* cannot contain forward slashes"
    ):
        MetaxyConfig(project="my/project")


def test_project_name_validation_double_underscore() -> None:
    """Test that project names cannot contain double underscores."""
    with pytest.raises(
        ValueError, match="project name .* cannot contain double underscores"
    ):
        MetaxyConfig(project="my__project")


def test_project_name_validation_invalid_chars() -> None:
    """Test that project names must be alphanumeric with underscores and hyphens."""
    with pytest.raises(
        ValueError,
        match="project name .* must contain only alphanumeric characters, underscores, and hyphens",
    ):
        MetaxyConfig(project="my.project")

    with pytest.raises(
        ValueError,
        match="project name .* must contain only alphanumeric characters, underscores, and hyphens",
    ):
        MetaxyConfig(project="my project")


def test_project_name_validation_valid_names() -> None:
    """Test that valid project names are accepted."""
    # Alphanumeric
    config1 = MetaxyConfig(project="myproject")
    assert config1.project == "myproject"

    # With underscores
    config2 = MetaxyConfig(project="my_project")
    assert config2.project == "my_project"

    # With hyphens
    config3 = MetaxyConfig(project="my-project")
    assert config3.project == "my-project"

    # Mixed
    config4 = MetaxyConfig(project="my_project-123")
    assert config4.project == "my_project-123"


def test_default_project_from_config() -> None:
    """Test that MetaxyConfig.get() returns default project when not initialized."""
    # Reset config to ensure clean state
    MetaxyConfig.reset()

    # Get default config - should have project='default'
    config = MetaxyConfig.get()
    assert config.project == "default"


def test_project_from_metaxy_toml(tmp_path: Path) -> None:
    """Test loading project from metaxy.toml."""
    # Create metaxy.toml with project setting
    config_file = tmp_path / "metaxy.toml"
    config_file.write_text(
        """
project = "my_metaxy_project"

[stores.dev]
type = "metaxy.metadata_store.InMemoryMetadataStore"
"""
    )

    # Load config
    config = MetaxyConfig.load(config_file)

    assert config.project == "my_metaxy_project"

    MetaxyConfig.reset()


def test_project_from_pyproject_toml(tmp_path: Path) -> None:
    """Test loading project from pyproject.toml [tool.metaxy] section."""
    config_file = tmp_path / "pyproject.toml"
    config_file.write_text(
        """
[project]
name = "test"

[tool.metaxy]
project = "pyproject_metaxy"

[tool.metaxy.stores.dev]
type = "metaxy.metadata_store.InMemoryMetadataStore"
"""
    )

    config = MetaxyConfig.load(config_file)

    assert config.project == "pyproject_metaxy"

    MetaxyConfig.reset()


def test_project_override_via_env_var(
    tmp_path: Path, monkeypatch: pytest.MonkeyPatch
) -> None:
    """Test that METAXY_PROJECT env var overrides config file."""
    # Create config file with one project
    config_file = tmp_path / "metaxy.toml"
    config_file.write_text(
        """
project = "file_project"

[stores.dev]
type = "metaxy.metadata_store.InMemoryMetadataStore"
"""
    )

    # Override via env var
    monkeypatch.setenv("METAXY_PROJECT", "env_project")

    config = MetaxyConfig.load(config_file)

    # Env var should win
    assert config.project == "env_project"

    MetaxyConfig.reset()


def test_multiple_features_same_project(snapshot: SnapshotAssertion) -> None:
    """Test that multiple features in the same graph share the same project."""
    config = MetaxyConfig(project="shared_project")
    MetaxyConfig.set(config)

    try:

        class Feature1(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["feature1"]),
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

        class Feature2(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["feature2"]),
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

        # Both features should have the same project
        assert Feature1.project == "shared_project"
        assert Feature2.project == "shared_project"
        assert Feature1.project == Feature2.project
        assert Feature1.project == snapshot

    finally:
        MetaxyConfig.reset()


def test_feature_project_persists_across_graph_operations(
    snapshot: SnapshotAssertion,
) -> None:
    """Test that project is preserved during graph operations like to_snapshot/from_snapshot."""
    config = MetaxyConfig(project="persist_project")
    MetaxyConfig.set(config)

    try:
        graph = FeatureGraph()

        with graph.use():
            # Define a test feature that will get the project from config
            class SnapshotTestFeature(
                Feature,
                spec=FeatureSpec(
                    key=FeatureKey(["snapshot", "test", "feature"]),
                    fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
                ),
            ):
                pass

        # Verify project before snapshot
        assert SnapshotTestFeature.project == "persist_project"

        # Create snapshot
        snapshot_data = graph.to_snapshot()

        # Verify snapshot contains project info
        feature_key_str = "snapshot/test/feature"
        assert feature_key_str in snapshot_data
        assert snapshot_data[feature_key_str]["project"] == "persist_project"

        # Verify that feature_tracking_version is in snapshot
        assert "feature_tracking_version" in snapshot_data[feature_key_str]

        # Store snapshot data for verification
        snapshot_project = snapshot_data[feature_key_str]["project"]
        snapshot_tracking = snapshot_data[feature_key_str]["feature_tracking_version"]

        # For a proper test, we would need the feature to be importable,
        # but we can at least verify the snapshot data is correct
        assert snapshot_project == "persist_project"
        assert snapshot_tracking is not None and len(snapshot_tracking) > 0

        # Verify the structure matches our snapshot
        assert {
            "project": snapshot_project,
            "tracking_version": snapshot_tracking[:8],
        } == snapshot

    finally:
        MetaxyConfig.reset()
