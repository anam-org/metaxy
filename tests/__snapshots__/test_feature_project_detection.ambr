# serializer version: 1
# name: test_feature_gets_project_from_config
  'test_project'
# ---
# name: test_feature_project_different_configs
  dict({
    'feature_a': 'project_a',
    'feature_b': 'project_b',
  })
# ---
# name: test_feature_project_persists_across_graph_operations
  dict({
    'project': 'persist_project',
    'tracking_version': '47e9595c',
  })
# ---
# name: test_multiple_features_same_project
  'shared_project'
# ---
